### PR TITLE
ctrl_vlan: add case to test 'ctrl_vlan' on linux and windows guest

### PR DESCRIPTION
**Pull request by sitoliu**
ID: 1585571
Signed-off-by: Sitong Liu <siliu@redhat.com>

### CODE REVIEW

File: qemu/tests/cfg/ctrl_vlan.cfg [hunk 1]
```diff
@@ -0,0 +1,19 @@
+- ctrl_vlan: install setup image_copy unattended_install.cdrom
```
Comment (YanhuiMa) [inline qemu/tests/cfg/ctrl_vlan.cfg]: You can delete content after colon. It doesn't work for avacodo.

File: qemu/tests/ctrl_vlan.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    3) Check vlan table in rx-filter information
+    4) If ctrl_vlan=on, do step 5-6
+    5) Set vlan in guest
+    6) Check vlan table in rx-filter information again
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    def verify_vlan_table(vm, expect_vlan=None):
```
Comment (YanhuiMa) [inline qemu/tests/ctrl_vlan.py]: No need to pass "vm" here.

File: qemu/tests/cfg/ctrl_vlan.cfg [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    type = ctrl_vlan
+    image_snapshot = yes
```
Comment (YanhuiMa) [inline qemu/tests/cfg/ctrl_vlan.cfg]: Just a quick question, why need snapshot here?

File: qemu/tests/ctrl_vlan.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        error_context.context("Check vlan table in rx-filter", logging.info)
+        query_cmd = "query-rx-filter name=%s" % vm.virtnet[0].device_id
```
Comment (YanhuiMa) [inline qemu/tests/ctrl_vlan.py]: Do you need to put query_cmd to cfg?

File: qemu/tests/cfg/ctrl_vlan.cfg (hunk 1 above)
Comment (sitoliu) [reply qemu/tests/cfg/ctrl_vlan.cfg]: Thanks for review, will remove it.

File: qemu/tests/cfg/ctrl_vlan.cfg (hunk 3 above)
Comment (sitoliu) [reply qemu/tests/cfg/ctrl_vlan.cfg]: To keep image clean after specific setting configure on guest.

File: qemu/tests/ctrl_vlan.py (hunk 2 above)
Comment (sitoliu) [reply qemu/tests/ctrl_vlan.py]: Thanks, will update it.

File: qemu/tests/ctrl_vlan.py (hunk 4 above)
Comment (sitoliu) [reply qemu/tests/ctrl_vlan.py]: Thanks, I think it is ok to put query_cmd here.

**Comment by YanhuiMa** [comment]
Acked-by: Yanhui Ma yama@redhat.com

**Comment by vivianQizhu** [comment]
@sitoliu Please fix CI failure.

File: qemu/tests/cfg/ctrl_vlan.cfg [hunk 5]
```diff
@@ -0,0 +1,20 @@
+- ctrl_vlan:
```
Comment (xiywang) [inline qemu/tests/cfg/ctrl_vlan.cfg]: is this patch depend on https://github.com/avocado-framework/avocado-vt/pull/1751?
if so, please add it in your first conversation.

File: qemu/tests/ctrl_vlan.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    def verify_vlan_table(expect_vlan=None):
+        error_context.context("Check vlan table in rx-filter", logging.info)
+        query_cmd = "query-rx-filter name=%s" % vm.virtnet[0].device_id
+        vlan_table = vm.monitor.send_args_cmd(query_cmd)[0].get("vlan-table")
+        if not expect_vlan:
+            vlan_table.sort()
+            if len(set(vlan_table)) == 4096 and vlan_table[0] == 0 and vlan_table[-1] == 4095:
```
Comment (xiywang) [inline qemu/tests/ctrl_vlan.py]: Have you tested this patch on RHEL.8.0 guest? The output before vlan set should be vlan-table[], no value anymore. Please check it again. Thanks.

Comment (xiywang) [reply qemu/tests/ctrl_vlan.py]: Moreover, after add a vlan to guest, vlan-table becomes [1, 0].
After remove the vlan, vlan-table becomes [0].
And if ctrl_vlan = off, vlan-table filled with [4095, 4094, 4093, ..., 0].

Comment (xiywang) [reply qemu/tests/ctrl_vlan.py]: Moreover, after add a vlan to guest, vlan-table becomes [1, 0].
After remove the vlan, vlan-table becomes [0].
And if ctrl_vlan = off, vlan-table filled with [4095, 4094, 4093, ..., 0].

As for rhel7.6, with ctrl_vlan=on
1). before set vlan in guest, "vlan-table": [0] (different from rhel8.0 output)
2). after set vlan in guest, "vlan-table": [1, 0] (the same as rhel8.0)
3). after remove vlan in guest, "vlan-table": [0] (the same as rhel8.0)
and with ctrl_vlan=off, the result is as the same of rhel8.0

Comment (sitoliu) [reply qemu/tests/ctrl_vlan.py]: @xiywang I writed the case according to the polarion case, if the behaviour is different on rhel8, I suggest to update the polarion case first.

Comment (xiywang) [reply qemu/tests/ctrl_vlan.py]: Updated the expected results in polarion.

File: qemu/tests/cfg/ctrl_vlan.cfg (hunk 5 above)
Comment (sitoliu) [reply qemu/tests/cfg/ctrl_vlan.cfg]: @xiywang Thanks, update it.

File: qemu/tests/ctrl_vlan.py (hunk 6 above)
Comment (sitoliu) [reply qemu/tests/ctrl_vlan.py]: @xiywang The result is weird for rhel8 guest, the vlan-table should not be `null list` when we already add a vlan for it, could you confirm if it is not a product bug?

Comment (xiywang) [reply qemu/tests/ctrl_vlan.py]: It isn't null. It's [1, 0] after add a vlan. Please check the expected result after step 5 in polarion.

Comment (sitoliu) [reply qemu/tests/ctrl_vlan.py]: @xiywang sorry, I mean before we add a vlan it should not be `non list`

**Comment by sitoliu** [comment]
Update it.
  (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.ctrl_vlan.ctrl_on: PASS (78.32 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.ctrl_vlan.ctrl_off: PASS (61.80 s)

**Pull request merged**